src/Data/Nat/Primality.agda

```diff
@@ -91,9 +301,10 @@ prime⇒¬composite {n@(suc (suc _))} n-prime (d , 2≤d , d<n , d∣n) =
 -- the ring theoretic definition of a prime element of the semiring ℕ.
 -- This is useful for proving many other theorems involving prime numbers.
 euclidsLemma : ∀ m n {p} → Prime p → p ∣ m * n → p ∣ m ⊎ p ∣ n
-euclidsLemma m n {p@(suc (suc _))} p-prime p∣m*n = result
+euclidsLemma m n {p} (prime pr) p∣m*n = result
   where
   open ∣-Reasoning
+  instance _ = nonTrivial⇒nonZero -- for NonZero p
```

Comment:
`nonTrivial⇒nonZero` should be an instance?

Reply:
Yes, ... but that causes (non-local!) problems, and I find this kind of thing such a headache to get right. Any ideas?

Reply:
I'm sorry, you'll have to give an idea of what the error messages are when you make it an instance?

Reply:
Here's an example:
```agda
/home/james/RESEARCH/Agda-stdlib-dev-v2.0-rc1/src/Data/Nat/DivMod.agda:309,22-40
No instance of type NonTrivial (n * o) was found in scope.
when checking that (*-comm o n) is a valid argument to a function
of type
{m = m₁ : ℕ} .⦃ _ = z : NonZero (o * n) ⦄
.⦃ _ = z₁ : NonZero (n * o) ⦄ →
o * n ≡ n * o → m₁ / (o * n) ≡ m₁ / (n * o)```

Reply:
The step at hand:
```agda
  n * m / (o * n) ≡⟨ /-congʳ {o = n * o} (*-comm o n) ⟩
  n * m / (n * o)
```
make *no* appeal to `NonTrivial`, and yet the instance means that the typechecker goes looking for one... very annoying! Is it a bug?

Reply:
Even when rewritten to make the instances more atomic, as:
```agda
m*n/o*n≡m/o : ∀ m n o .⦃ _ : NonZero n ⦄ ⦃ _ : NonZero o ⦄ →
              let instance _ = m*n≢0 o n in 
              m * n / (o * n) ≡ m / o
m*n/o*n≡m/o m n o = begin-equality
  m * n / (o * n) ≡⟨ /-congˡ (*-comm m n) ⟩
  n * m / (o * n) ≡⟨ /-congʳ {o = n * o} (*-comm o n) ⟩
  n * m / (n * o) ≡⟨ m*n/m*o≡n/o n m o ⟩
  m / o           ∎
  where instance _ = m*n≢0 n o; instance _ = m*n≢0 o n
```

Reply:
Having now pushed to the bitter end of #2182 in its simplification of `NonZero` instance reasoning in `Data.Nat.DivMod` , making the above `nonTrivial⇒nonZero` into an `instance` now breaks the lemma `m*n/m*o≡n/o` instead ... so something is going on, but I'm struggling to be able to precisely nail it down.